### ISSUE TITLE
Provide flags to configure metal compilation

### ISSUE THREAD

**Issue by schwa** [apple/swift-package-manager#5823]
### Description

It seems SPM now magically compiles .metal files into a default.metallib (see https://github.com/apple/swift-package-manager/issues/5822)

Unfortunately there's no documented way to specify compilation flags to the metal compiler.

This means it is impossible to debug shaders using the Xcode Metal Debugger. When we try we get this error:

> Unable to create shader debug session
> 
> Source is unavailable
> 
> Under the target's Build Settings, ensure the Metal Compiler Build Options produces debugging information and includes source code.
> 
> If building with the 'metal' command line tool, include the options '-gline-tables-only' and '-frecord-sources'.

I'd like to see .metalFlags or similar available to SPM

**Comment by neonichu** [comment]
Generally, I'm a bit skeptical about adding manifest API for Xcode specific file types. For this issue in particular, I think Xcode should just automatically do the right thing and pass debug-related flags (there's an existing bug tracking that request rdar://79153205).

cc @abertelrud

**Comment by schwa** [comment]
Boiling the problem down and we're left with "we cannot debug Metal shaders from SPM Packages"

I'm not concerned how that's solved. But right now it means folks can't really use SPM & metal together (see https://twitter.com/warrenm/status/1582039539123486721).

I've been trying to get it working and have resorted to using my own SPM Plugin to compile .metal https://github.com/schwa/MetalCompilerPlugin - which isn't ideal as I can't overwrite default.metallib and have to build my "debug.metallib" side by side with it.

That said I can definitely see a need to send metal compilation flags to metal from SPM - being able to configure -D macros for example seem very useful.

**Comment by schwa** [comment]
(As background: One technique folks have been using as well is to put their metal files in a C target, with headers in ./include and a module map to make the headers available to swift.)

**Comment by wtholliday** [comment]
> Generally, I'm a bit skeptical about adding manifest API for Xcode specific file types. For this issue in particular, I think Xcode should just automatically do the right thing and pass debug-related flags (there's an existing bug tracking that request rdar://79153205).
> 
> cc @abertelrud

Yes, Xcode should just do automatically the right thing. But it doesn't. We've been hacking around this stupid problem for years now. Throw us a bone.

**Comment by schwa** [comment]
It's been almost a year - is there still no way to configure how SPM compiles metal code into .metallibs?

**Comment by neonichu** [comment]
Looks like rdar://79153205 is still unresolved.

**Comment by code-per-day** [comment]
Checking the pulse on this issue, is there still no way to debug metal shaders with SwiftPM?

**Comment by wtholliday** [comment]
AFAIK, no. I've abandoned SPM for most of my code because of this. It's just too much of a headache.

**Comment by schwa** [comment]
I _am_ debugging metal shaders using the SPM plugin I mentioned earlier (https://github.com/schwa/MetalCompilerPlugin). It's not a great solution (development-only plugins suck in general) but it does allow me to debug shaders.

I'd still much prefer a proper solution from Apple but this definitely does not seem overly important to them.
